refactor and fix test cases for mrw

ArcturusZhang opened the pull request:
This PR contains two parts: refactoring and fixing test issues.

## refactors
1. We are building up strings to get the expected result. I changed them to use JsonSerializer along with anonymous objects (see [this comment](https://github.com/Azure/azure-sdk-for-net/pull/41055/files#r1444158556)).
2. since the purpose of the "roundtrip" test cases, are validating if the serialization and deserialization could be roundtrip, where roundtrip means if the JSON is the same, and the JSON string, I believe does not have to be identical. Therefore I changed the way of validating the result (see [this comment](https://github.com/Azure/azure-sdk-for-net/pull/41055/files#r1444158879)). The test cases here might be OK unchanged (because they previously could pass), but since the generator side, we generate those models and the sequence of writing properties is slightly different, I have to change the sequence of the "expected result" a lot, which is too cumbersome to maintain therefore I change the way of validating results.

## fixes for test cases
1. The model `AvailabilitySetData` is not fully implemented, therefore its roundtrip test case is not fully passed.
2. The type of exception should align - in real implementation we will never throw `InvalidOperationException`.

ArcturusZhang committed:
refactor and fix test cases for mrw

ArcturusZhang commented on sdk/core/System.ClientModel/tests/ModelReaderWriter/Models/AvailabilitySetDataTests.cs:
```diff
@@ -17,17 +18,49 @@ internal class AvailabilitySetDataTests : ModelJsonTests<AvailabilitySetData>
-            return expectedSerializedString; ;
+            object obj = format switch
+            {
+                "J" => new
```
I believe using anonymous objects here are making this more clear comparing with building up literal strings

ArcturusZhang commented on sdk/core/System.ClientModel/tests/ModelReaderWriter/Models/ModelTests.cs:
```diff
@@ -71,12 +72,82 @@ protected void RoundTripTest(string format, RoundTripStrategy<T> strategy)
 
-            Assert.That(roundTrip, Is.EqualTo(expectedSerializedString));
+            // we validate those are equivalent element, representing the same json object (ignoring the spaces and orders, etc)
+            AssertJsonEquivalency(expectedSerializedString, roundTrip);
```
It is not quite flexible if we are just comparing the string results because depending on the implementation of the serialization, the same model could result in different json strings but they represent the same JSON object.

ArcturusZhang commented on sdk/core/System.ClientModel/tests/ModelReaderWriter/Models/ModelTests.cs:
```diff
@@ -173,18 +244,18 @@ public void ThrowsIfWireIsNotJson()
             if (ModelInstance is IJsonModel<T> jsonModel && IsXmlWireFormat)
             {
-                Assert.Throws<InvalidOperationException>(() => jsonModel.Write(new Utf8JsonWriter(new MemoryStream()), _wireOptions));
+                Assert.Throws<FormatException>(() => jsonModel.Write(new Utf8JsonWriter(new MemoryStream()), _wireOptions));
```
We always throw FormatException when the format is not expected (see the above test case)
This case could previously pass because we always has the flag `IsXmlWireFormat` false.

ArcturusZhang commented on sdk/core/System.ClientModel/tests/client/ModelReaderWriter/ServiceModels/AvailabilitySetData.Serialization.cs:
```diff
@@ -85,6 +85,21 @@ private void Serialize(Utf8JsonWriter writer, ModelReaderWriterOptions options)
                 JsonSerializer.Serialize(writer, ProximityPlacementGroup);
             }
             writer.WriteEndObject();
+            if (options.Format != "W" && _serializedAdditionalRawData != null)
```
this model previously are not implemented with the additional raw data during serialization and deserialization. Here we added it.

azure-sdk commented:
**API change check**

API changes are not detected in this pull request.

JoshLove-msft commented on sdk/core/System.ClientModel/tests/ModelReaderWriter/Models/ModelTests.cs:
```diff
@@ -71,12 +72,82 @@ protected void RoundTripTest(string format, RoundTripStrategy<T> strategy)
+                    AssertJsonArrayEquivalency(expected, result);
+                    break;
+                default:
+                    Assert.AreEqual(expected.ToString(), result.ToString());
```
Would this be correct for something like a bool or int? It seems that if the extected value is bool with value true, but the actual is the string "true", this would still return that they are equivalent.

JoshLove-msft replied:
Nevermind, I see that there would be different ValueKinds in that case and the assertion would fail on L94.

ArcturusZhang replied:
I tried if we could just assert they are equal without toString, but it does not work.
Maybe `GetRawText` is a better solution here?

annelo-msft reviewed (changes requested):
It makes me very nervous to change tests to GA Core components.  The reason for this is that the tests are validating behaviors we have a contract with customers that we will not make breaking changes to, and it's not clear to me how we can be certain that we won't accidentally change the tests in a way that allows us to introduce regressions for things that were previously validated.  I'd like to work with the Core v-team to establish best practices on this topic before moving forward with changes to tests in either Azure.Core or System.ClientModel test projects.  My proposal is that we adopt a policy where we don't make changes to tests unless one of two cases holds: 1) We find a bug in the test or 2) we decide to change the functionality the test is validating.  

If we think we have a set of tests that needs to change with some degree of regularity, I'd love it if we could take a step back to ask why.  Specifically:
1. What *unit* test coverage do we have of the Core component?  Is it complete?
2. If the tests that need to change are *scenario* tests, should they live in Core at all?

There is a philosophy of tests that says every test should validate one specific behavior of a component.  If we could clarify the type and behaviors of the Core components that these tests are testing, I think it might make it easier to understand the risks/tradeoff around changing tests of ModelReaderWriter.

jsquire reviewed (dismissed):
I'd like to understand this before it moves forward.  Core/ClientModel is consumed by the generator and should not be dependent on it - we want the dependency graph to flow in one direction.  Otherwise, we end up right back in the circular dependency relationship again, and that's not a good place to be.

I'm also not clear on what is changing here and why.  Fundamentally, the tests in Core should be stable and NOT regenerate based on changes elsewhere.  If we have data that is independent of the asserts being done that we want to feed in, I've no objection - but that should be an external data source not embedded in the tests.

It seems to me that we may need to split this out.  Core has a known stable set of data that corresponds to mock models for test purposes.  These get updated when we add functionality to the Core component.   The generator (or generated libraries?) have a set that validates the specific models being generated, which acts as a validation of the generated code - not the Core functionality.


**UPDATE:**  This comment was based on a misunderstanding that this was a generated change and would be so going forward.  As this is not that, removing my blocker.   Since I've heard that we're considering generated changes like this in the future, leaving my comment here for future consideration.

ArcturusZhang commented on sdk/core/System.ClientModel/tests/ModelReaderWriter/Models/AvailabilitySetDataTests.cs:
```diff
@@ -17,17 +18,49 @@ internal class AvailabilitySetDataTests : ModelJsonTests<AvailabilitySetData>
-            if (format == "J")
-                expectedSerializedString += "\"name\":\"testAS-3375\",\"id\":\"/subscriptions/e37510d7-33b6-4676-886f-ee75bcc01871/resourceGroups/testRG-6497/providers/Microsoft.Compute/availabilitySets/testAS-3375\",\"type\":\"Microsoft.Compute/availabilitySets\",";
-            expectedSerializedString += "\"sku\":{\"name\":\"Classic\"";
-            //if (!ignoreAdditionalProperties)
-            //    expectedSerializedString += ",\"extraSku\":\"extraSku\"";
-            expectedSerializedString += "},\"tags\":{\"key\":\"value\"},\"location\":\"eastus\",\"properties\":{\"platformUpdateDomainCount\":5,\"platformFaultDomainCount\":3}";
-            //if (!ignoreAdditionalProperties)
-            //    expectedSerializedString += ",\"extraRoot\":\"extraRoot\"";
```
these two properties exist in the json file of this model, but we are not expecting them in the result?
This is not true unless this is on purpose.

ArcturusZhang commented:
> It makes me very nervous to change tests to GA Core components. The reason for this is that the tests are validating behaviors we have a contract with customers that we will not make breaking changes to, and it's not clear to me how we can be certain that we won't accidentally change the tests in a way that allows us to introduce regressions for things that were previously validated. I'd like to work with the Core v-team to establish best practices on this topic before moving forward with changes to tests in either Azure.Core or System.ClientModel test projects. My proposal is that we adopt a policy where we don't make changes to tests unless one of two cases holds: 1) We find a bug in the test or 2) we decide to change the functionality the test is validating.
> 
> If we think we have a set of tests that needs to change with some degree of regularity, I'd love it if we could take a step back to ask why. Specifically:
> 
> 1. What _unit_ test coverage do we have of the Core component?  Is it complete?
> 2. If the tests that need to change are _scenario_ tests, should they live in Core at all?
> 
> There is a philosophy of tests that says every test should validate one specific behavior of a component. If we could clarify the type and behaviors of the Core components that these tests are testing, I think it might make it easier to understand the risks/tradeoff around changing tests of ModelReaderWriter.

I cannot answer your question here in general. But this PR contains two parts: refactor the test case code and fix some test cases.

I was introducing some of the test cases here to the generator part to make sure the generator is generating those models that could meet the requirement of MRW, during which I noticed some of the test cases are not quite simple to maintain.
1. We are building up strings to get the expected result. I changed them to use `JsonSerializer` along with anonymous objects (see [this comment](https://github.com/Azure/azure-sdk-for-net/pull/41055/files#r1444158556)).
2. since the purpose of the "roundtrip" test cases, are validating if the serialization and deserialization could be roundtrip, where roundtrip means if the JSON is the same, and the JSON string, I believe does not have to be identical. Therefore I changed the way of validating the result (see [this comment](https://github.com/Azure/azure-sdk-for-net/pull/41055/files#r1444158879)). The test cases here might be OK unchanged (because they previously could pass), but since the generator side, we generate those models and the sequence of writing properties is slightly different, I have to change the sequence of the "expected result" a lot, which is too cumbersome to maintain therefore I change the way of validating results.

Above are the "refactors" introduced by this PR.

The fixes I want to introduce in this PR is for this case:
1. The `AvailabilitySetData` model is not fully implemented therefore we are not fully validating the roundtrip case for it: https://github.com/Azure/azure-sdk-for-net/pull/41055/files#r1446813585
2. This might be depending on how the interfaces are implemented but we should align with other test cases and the implementation here: https://github.com/Azure/azure-sdk-for-net/pull/41055/files#r1444159093

I also think, all those test cases I am changing, the serialization and deserialization tests and the roundtrip validations, they depend on how the model implements the interfaces we introduced in this library. Would it be a test issue if the model for test is not properly implemented as the purpose indicates? For instance, the implementation of `AvailabilitySetData` model is not implemented with the `serializedAddtionalRawData` therefore it could never be roundtrip in a row of deserializing and then serialize.

ArcturusZhang committed:
Merge branch 'main' into fix-and-refactor-mrw-tests
